### ISSUE TITLE
[calcite-table] Refactor component to improve compatibility with lit-html

### ISSUE THREAD

**Issue by maxpatiiuk** [Esri/calcite-design-system#10495]
### Description

Post-codemod, most Calcite components work with no or minimal changes.
However, `<calcite-table>` may require some changes because of the way lit-html works.

lit-html converts each template to a `<template>` element behind the scenes. That means each template needs to be standalone piece of valid HTML.

That is causing issues in `<calcite-table>` - since the table is rendered in pieces, some templates are not valid standalone (tables have strict rules about what elements can be inside of them), and so the browser moves the "invalid" elements out of the table.
 
[Code sample](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxuQGN1c3RvbUVsZW1lbnQoJ3NpbXBsZS1ncmVldGluZycpXG5leHBvcnQgY2xhc3MgU2ltcGxlR3JlZXRpbmcgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IGNzc2A6aG9zdCB7IGRpc3BsYXk6IGNvbnRlbnRzOyB9YDtcbiAgcmVuZGVyKCkge1xuICAgIHJldHVybiBbXG4gICAgICAgIHRoaXMucmVuZGVySGVhZGVyKCksXG4gICAgICAgIHRoaXMucmVuZGVyQm9keSgpLFxuICAgICAgICB0aGlzLnJlbmRlckZvb3RlcigpLFxuICAgIF07XG4gIH1cbiAgcmVuZGVySGVhZGVyKCkge1xuICAgIHJldHVybiBodG1sIGA8dGhlYWQ-PHNsb3QgbmFtZT1cImhlYWRlclwiIC8-PC90aGVhZD5gO1xuICB9XG4gIHJlbmRlckJvZHkoKSB7XG4gICAgcmV0dXJuIGh0bWwgYDx0Ym9keT48c2xvdCAvPjwvdGJvZHk-YDtcbiAgfVxuICByZW5kZXJGb290ZXIoKSB7XG4gICAgcmV0dXJuIGh0bWwgYDx0Zm9vdD48c2xvdCBuYW1lPVwiZm9vdGVyXCIgLz48L3Rmb290PmA7XG4gIH1cbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9zaW1wbGUtZ3JlZXRpbmcuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8dGFibGU-XG4gICAgPHNpbXBsZS1ncmVldGluZz5cbiAgICAgIDx0ciBzbG90PVwiaGVhZGVyXCI-PHRoPkhlYWRlcjwvdGg-PC90cj5cbiAgICAgIDx0cj48dGQ-Q2VsbDwvdGQ-PC90cj5cbiAgICA8L3NpbXBsZS1ncmVldGluZz5cbiAgPC90YWJsZT5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4zLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl40LjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMy4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0) (take a look in devtools at what does the rendered html look like)
 
The [solution proposed by Lit](https://discord.com/channels/1012791295170859069/1047015641225371718/threads/1168736743432917132) is to use div[role="cell"] and etc instead of td inside the calcite-table components.
then with css apply display:table-cell;
Example: https://lit.dev/playground/#gist=0b495facaf08c1f990e9879f54405fc8
 
The other options might be to rendering more parts of the table as a single lit-html template, or doing imperative rendering in table components.
 
As far as I am aware, other component are not subject to this issue - it's just table that has these restrictions.

### Proposed Advantages

calcite-table needs to be refactored a bit to render correctly in lit-html.
Until then, the browser may move what it sees as "HTML element with incorrect placement" out of the table, breaking the layout:

![Screenshot 2024-10-07 at 12 42 50](https://github.com/user-attachments/assets/76a36697-3423-4c61-97dd-edd5dfdccee3)

### Which Component

calcite-table

### Calcite package

- [X] @esri/calcite-components
- [ ] @esri/calcite-components-angular
- [ ] @esri/calcite-components-react
- [ ] @esri/calcite-design-tokens
- [ ] @esri/eslint-plugin-calcite-components

**Comment by macandcheese** [comment]
#6646 will provide imperative rendering, but I know there is value for a lot of our "dev summit users users" simple / use cases to stand up a static table declaratively. If that can be preserved in some way, that would be nice for those cases - alongside the more advanced functionality of course.

**Comment by maxpatiiuk** [comment]
> If that can be preserved in some way, that would be nice for those cases

Agree. More research needs to be done, but it appears that the required changes would be done inside the calcite-table code, rather than affecting outside usage.

e.g. if it worked in Stencil, it should work in Lit. The only thing that changed is the rendering engine used by calcite-table.

**Comment by driskull** [comment]
Is this an issue with calcite-list? It places divs around <tr> etc.

**Comment by maxpatiiuk** [comment]
> Is this an issue with calcite-list? It places divs around <tr> etc.

You are correct, that appears to be the case. I did not realize calcite-list was using a table underneath.
I will edit the issue title to mention both.

**Comment by driskull** [comment]
> so the browser moves the "invalid" elements out of the table.

Is this really the case? Wouldn't this be occurring now with these components? This seems like some kind of lit thing not a browser thing.

**Comment by driskull** [comment]
I think it may be more than just converting to a div with a role. The colSpan/rowSpan styling might be challenging here.

I'd also be concerned with how this affects a11y. If a11y does have an issue with a component between a tr/td then maybe its worth it. Otherwise its a lot to work around.

**Comment by macandcheese** [comment]
I tried the role approach and pushed an example `macandcheese/lit-table-demo` if you want to look - I see a similar result visually, though I could have missed a couple styles in the changes. You  do of course lose "native" colSpan / rowSpan support as mentioned.

**Comment by driskull** [comment]
@maxpatiiuk does it also have an issue with `<li>` elements that are not within a `<ul>/<ol>`? Because we have some occurrences of that as well.

**Comment by maxpatiiuk** [comment]
> Edit - oops, of course this is the Stencil version, so you'd need to run the codemod on that branch to see the "true" result I suppose.

I have a Lumina branch in this PR - feel free to experiment on top of it:

https://github.com/Esri/calcite-design-system/pull/10482

But you can also create a small lit app (not even lumina) and play around with how it renders a table - see for example this Lit playground ["Code sample"](https://github.com/Esri/calcite-design-system/issues/10495#:~:text=Code%20sample%20(take%20a%20look%20in%20devtools%20at%20what%20does%20the%20rendered%20html%20look%20like))

**Comment by maxpatiiuk** [comment]
Lit also has static templates where you can create template from dynamic parts - https://lit.dev/docs/templates/expressions/#static-expressions

You could have renderTHead return an HTML string, that is inlined inside the render() method's template.
Not saying this is ideal DX, but there are options so need to investigate them

**Comment by driskull** [comment]
> If this is the core of the issue, then maybe it's just about collapsing multiple functional components inside calcite-table into one

I think this is the issue but this wouldn't work since the table rows/cells are slotted in, so combining them with the table is not a viable option.

> the issue is about the divs in between, then that I ask you to research why this wasn't an issue in Stencil and how we can workaround.

This is also an issue since HTML table requires certain direct children. However, stencil doesn't stop you from doing this and it doesn't seem like screen readers care if the HTML isn't valid. It seems like lit does stop you potentially because of the HTML template being used.

**Comment by macandcheese** [comment]
I checked out the Lumina branch - it seems like the Table seems to render mostly correctly in some cases (there are some styling things here and there, but the structure / layout is accurate). 

However the cases that appear broken (as shown in screenshot) - we are programmatically rendering additional cells on behalf of the user - this occurs for `numbered` and when `selectionMode` is set. Maybe we can play with the new lifecycle methods and see if rendering these at a different time / in a different matter works.

**Comment by driskull** [comment]
I have a PR open to fix the list component.

We may need to identify other components that may be affected with `ul > li` where the `li` isn't a direct child of `ul`.

This is one of those quirks with web components where the parent/child can't be separated by a shadow DOM.

**Comment by driskull** [comment]
I think the following components will need updating as well:

- combobox/combobox-item
- menu/menu-item

**Comment by driskull** [comment]
@jcfranco do we want to update this issue to include combobox and menu? I think those are the only ones using `ol > li` inappropriately.

**Comment by driskull** [comment]
Ok maybe tables are the only ones with immediate issues. We should have a followup issue to refactor the `ul > li` ones post 3.0

**Comment by driskull** [comment]
List PR has been installed so it should be good now.

**Comment by DitwanP** [comment]
🍡 Verified
